File: 백준/Gold/2169. 로봇 조종하기/로봇 조종하기.py
import copy
n, m = map(int, input().split())
data = [list(map(int, input().split())) for _ in range(n)]

dp = [[0]*m for _ in range(n)]
# 첫행은 누적으로 dp채워주기
dp[0] = copy.deepcopy(data[0])

for i in range(1, m):
    dp[0][i] += dp[0][i-1]

for i in range(1, n):
    tmp1 = [0]*m  # 왼쪽 > 오른쪽 최대값
    tmp2 = [0]*m  # 왼쪽 < 오른쪽 최대값

    for j in range(m):

        # 첫번째 칸은 그냥 갱신
        if j == 0:
            tmp1[j] = data[i][j] + dp[i-1][j]
            tmp2[m-1-j] = data[i][m-1-j] + dp[i-1][m-1-j]
            continue

        tmp1[j] = data[i][j] + max(dp[i-1][j], tmp1[j-1])
        tmp2[m-1-j] = data[i][m-1-j] + max(dp[i-1][m-1-j], tmp2[m-j])

    # tmp1과 tmp2 중 최대값을 dp에 갱신
    tmp = [max(tmp1[i], tmp2[i]) for i in range(m)]
    dp[i] = copy.deepcopy(tmp)

print(dp[n-1][m-1])

File: 백준/Gold/2169. 로봇 조종하기/로봇 조종하기.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin', 'utf8').trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const data = input.slice(1).map(line => line.split(' ').map(Number));

// DP 테이블 초기화
const dp = Array.from({ length: n }, () => Array(m).fill(0));
dp[0] = [...data[0]];

// 첫 번째 행을 누적으로 채우기
for (let i = 1; i < m; i++) {
    dp[0][i] += dp[0][i - 1];
}

for (let i = 1; i < n; i++) {
    let tmp1 = Array(m).fill(0); // 왼쪽 > 오른쪽 최대값
    let tmp2 = Array(m).fill(0); // 오른쪽 > 왼쪽 최대값

    for (let j = 0; j < m; j++) {
        if (j === 0) {
            tmp1[j] = data[i][j] + dp[i - 1][j];
            tmp2[m - 1 - j] = data[i][m - 1 - j] + dp[i - 1][m - 1 - j];
            continue;
        }
        tmp1[j] = data[i][j] + Math.max(dp[i - 1][j], tmp1[j - 1]);
        tmp2[m - 1 - j] = data[i][m - 1 - j] + Math.max(dp[i - 1][m - 1 - j], tmp2[m - j]);
    }

    // tmp1과 tmp2 중 최대값을 dp에 갱신
    dp[i] = tmp1.map((val, idx) => Math.max(val, tmp2[idx]));
}

console.log(dp[n - 1][m - 1]);
